unison/lockmanager.go

```diff
@@ -0,0 +1,427 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package unison
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/elastic/go-concert"
+	"github.com/elastic/go-concert/atomic"
+)
+
+// LockManager gives access to a set of Locks by name.  The lock manager can
+// forcefully unlock a lock. Routines using a manged lock can use the
+// LockSession to list for special Lock events.
+//
+// The zero value of LockManager is directly usable, but a LockManager must no
+// be copied by value.
+type LockManager struct {
+	initOnce sync.Once
+	mu       sync.Mutex
+	table    map[string]*lockEntry
+}
+
+// ManagedLock is a mutex like structure that is maanged by the LockManager.
+// A managed lock can loose it's lock lease at any time. The LockX methods
+// return a LockSession that can be used to listen for the current Locks state
+// changes.
+//
+// The lock will automatically be released in case the ManagedLock is garbage
+// collected. One should not rely on this behavior, but releaseing a zombie
+// lock guarantees that other routines might eventually be able to make
+// progress in case of fatal errors.
+type ManagedLock struct {
+	key     string
+	manager *LockManager
+	session *LockSession
+	entry   *lockEntry
+}
+
+// LockOption is used to pass additonal settings to all LockX methods of the ManagedLock.
+type LockOption interface {
+	apply(l *ManagedLock)
+}
+
+// LockSession provides signal with the current lock state. Lock sessions must
+// not be reused, as each Lock operation returns a new Session object.
+type LockSession struct {
+	isLocked             atomic.Bool
+	done, unlocked, lost *concert.OnceSignaler
+}
+
+// WithSignalCallbacks is a LockOption that configures additional callbacks to be executed on lock session state changes.
+type WithSignalCallbacks struct {
+	Done     func()
+	Unlocked func()
+	Lost     func()
+}
```

Comment:
Maybe `Lock` callback would be nice to?

Reply:
The callbacks are per LockSession. The LockSession is alive from the moment any of the `LockX` functions did succeed, until after the lock has been unlocked. The is no real need for a 'Locked' callback The `Lost` signal can be triggered before `Unlocked`. The `Done` signal is triggered the moment `Lost` or `Unlocked` is triggered. Hm... maybe I should add some better docs.

Reply:
Sound good. I was surprised by all these callbacks and I've just said of I think I am missing one.